testsuite/tests/typing-poly/poly.ml

```diff
@@ -1949,3 +1937,54 @@ Line 2, characters 16-24:
 Error: This field value has type 'b option ref which is less general than
          'a. 'a option ref
 |}]
+
+
+(** #12210: turn row variable into univars under Ptyp_poly: *)
+
+let simple: 'a. 'a -> [> `X of 'a ] -> 'a = fun default ->
+  function
+  | `X x -> x
+  | _ -> default
+[%%expect {|
+val simple : 'a -> [> `X of 'a ] -> 'a = <fun>
+|}]
+
+type 'a w = Int: int w
+let locally_abstract: type a. a w -> [> `X of a ] -> a = fun Int ->
+  function
+  | `X x -> x
+  | _ -> 0
+[%%expect {|
+type 'a w = Int : int w
+val locally_abstract : 'a w -> [> `X of 'a ] -> 'a = <fun>
+|}]
+
+let nested: 'a.
+  <m: 'b.
+        <n:'irr.
+             ('irr -> unit) * ([> `X of 'a | `Y of 'b ] -> 'a)
+        >
+  > -> 'a  =
+  fun o -> (snd o#m#n) (`Y 0)
+[%%expect {|
+val nested :
+  < m : 'c 'b.
+          < n : 'irr.
+                  ('irr -> unit) * (([> `X of 'a | `Y of 'b ] as 'c) -> 'a) > > ->
+  'a = <fun>
+|}]
+
+
+let fail: 'a . 'a -> [> `X of 'a ] -> 'a = fun x y ->
+  match y with
+  | `Y -> x
+  | `X x -> x
```

Comment:
This example is now rejected, but it was accepted in older versions of OCaml. Let me explain what is going on, in case users hit their issue with their code that starts being rejected after the present PR is merged.

In the 5.0 toplevel (without the present PR):

```ocaml
# let fail: 'a . 'a -> [> `X of 'a ] -> 'a = fun x y ->
    match y with
    | `Y -> x
    | `X x -> x;;
val fail : 'a -> [< `X of 'a | `Y > `X ] -> 'a = <fun>
```

What was going on before is that the type annotation `[> `X of 'a ]` was too general, more general than what the implementation does, and the implementation was unifying the row variable to something more specific during type-checking of the implementation of the function. This is similar to what happens on `let f : 'a -> 'a = fun x -> x + 1`, which is (surprisingly?) accepted by OCaml, which happily specializes `'a` into `int`.

On the other hand, `let f : 'a . 'a -> 'a = fun x -> x + 1` is rejected, because `'a . ...` enforces that `'a` remains polymorphic. This is exactly why `fail` is rejected by this PR: now any open polymorphic variant that mentions an explicitly-polymorphic variable (the `'a` in ``[> `X of 'a ]``) is forced to remain polymorphic, so type-checking the function definition now fails.

The fix is to write directly the "right" annotation for this program from the start, not a more general version that needs to be un-generalized during type inference:

```ocaml
let fail: 'a . 'a -> [< `X of 'a | `Y ] -> 'a = fun x y ->
    match y with
    | `Y -> x
    | `X x -> x;;
```

( @Octachron maybe you could add this "fixed" version in the testsuite?)